src/realm/object-store/impl/collection_notifier.cpp

```diff
@@ -61,32 +63,64 @@ CollectionNotifier::get_modification_checker(TransactionChangeInfo const& info,
     return DeepChangeChecker(info, *root_table, m_related_tables);
 }
 
-void DeepChangeChecker::find_related_tables(std::vector<RelatedTable>& out, Table const& table)
+void DeepChangeChecker::find_related_tables(DeepChangeChecker::RelatedTables& out, Table const& table)
 {
-    auto table_key = table.get_key();
-    if (any_of(begin(out), end(out), [=](auto& tbl) {
-            return tbl.table_key == table_key;
-        }))
-        return;
+    struct ColKeyHasher {
+        std::size_t operator()(const ColKey& key) const
+        {
+            return std::hash<decltype(ColKey::value)>{}(key.value);
+        }
+    };
+    struct LinkInfo {
+        std::unordered_set<ColKey, ColKeyHasher> link_columns;
```

Comment:
A vector which is deduplicated after looping over the tables with sort/unique is likely to perform better than a set since insertion and lookup don't need to be mixed.